File: textbook/python/05-ml/08_hyperparameter_search.py
# Copyright 2026 Terrene Foundation
# SPDX-License-Identifier: Apache-2.0
"""
# ════════════════════════════════════════════════════════════════════════
# Kailash SDK Textbook — ML / HyperparameterSearch
# ════════════════════════════════════════════════════════════════════════
# OBJECTIVE: Optimize hyperparameters using grid, random, and Bayesian
#            search strategies.  Covers SearchSpace, ParamDistribution,
#            SearchConfig, and the search() entry point.
# LEVEL: Intermediate
# PARITY: Python-only (kailash-ml has no Rust counterpart)
# VALIDATES: HyperparameterSearch, SearchSpace, ParamDistribution,
#            SearchConfig, TrialResult, SearchResult — search(),
#            sample_grid(), sample_random()
#
# Run: uv run python textbook/python/05-ml/08_hyperparameter_search.py
# ════════════════════════════════════════════════════════════════════════
"""
from __future__ import annotations

import asyncio
import tempfile

import polars as pl

from kailash.db.connection import ConnectionManager
from kailash_ml import FeatureField, FeatureSchema
from kailash_ml.engines.feature_store import FeatureStore
from kailash_ml.engines.hyperparameter_search import (
    HyperparameterSearch,
    ParamDistribution,
    SearchConfig,
    SearchResult,
    SearchSpace,
    TrialResult,
)
from kailash_ml.engines.model_registry import LocalFileArtifactStore, ModelRegistry
from kailash_ml.engines.training_pipeline import EvalSpec, ModelSpec, TrainingPipeline


async def main() -> None:
    # ── 1. Set up infrastructure ────────────────────────────────────────

    conn = ConnectionManager("sqlite:///:memory:")
    await conn.initialize()

    with tempfile.TemporaryDirectory() as artifact_dir:
        artifact_store = LocalFileArtifactStore(artifact_dir)
        registry = ModelRegistry(conn, artifact_store)
        fs = FeatureStore(conn)
        await fs.initialize()

        pipeline = TrainingPipeline(feature_store=fs, registry=registry)
        search = HyperparameterSearch(pipeline)
        assert isinstance(search, HyperparameterSearch)

        # ── 2. Create synthetic data ────────────────────────────────────

        n_samples = 150
        df = pl.DataFrame(
            {
                "entity_id": [f"e{i}" for i in range(n_samples)],
                "feat_a": [float(i % 8) for i in range(n_samples)],
                "feat_b": [float(i * 0.3) for i in range(n_samples)],
                "target": [i % 2 for i in range(n_samples)],
            }
        )

        schema = FeatureSchema(
            name="hp_search_demo",
            features=[
                FeatureField(name="feat_a", dtype="float64"),
                FeatureField(name="feat_b", dtype="float64"),
            ],
            entity_id_column="entity_id",
        )

        # ── 3. Define a SearchSpace ─────────────────────────────────────
        # ParamDistribution supports: uniform, log_uniform, int_uniform,
        # categorical.

        space = SearchSpace(
            params=[
                ParamDistribution(
                    name="n_estimators",
                    type="int_uniform",
                    low=5,
                    high=30,
                ),
                ParamDistribution(
                    name="max_depth",
                    type="categorical",
                    choices=[3, 5, 7],
                ),
            ]
        )

        # ── 4. SearchSpace.sample_grid() — exhaustive grid ──────────────
        # For int_uniform with low=5, high=30 this generates all integers.
        # For categorical it uses all choices.

        grid = space.sample_grid()
        assert isinstance(grid, list)
        assert len(grid) > 0
        # Each element is a dict of param_name -> value
        assert "n_estimators" in grid[0]
        assert "max_depth" in grid[0]

        # ── 5. SearchSpace.sample_random() — random samples ──────────────

        random_samples = space.sample_random(n=10)
        assert isinstance(random_samples, list)
        assert len(random_samples) == 10
        for sample in random_samples:
            assert "n_estimators" in sample
            assert "max_depth" in sample
            # int_uniform produces integers
            assert isinstance(sample["n_estimators"], int)
            assert 5 <= sample["n_estimators"] <= 30
            # categorical draws from choices
            assert sample["max_depth"] in [3, 5, 7]

        # ── 6. ParamDistribution types ──────────────────────────────────

        # uniform: continuous float between low and high
        uniform_param = ParamDistribution("lr", "uniform", low=0.001, high=0.1)

        # log_uniform: log-scale sampling (useful for learning rates)
        log_param = ParamDistribution("lr_log", "log_uniform", low=0.001, high=0.1)

        mixed_space = SearchSpace(params=[uniform_param, log_param])
        mixed_samples = mixed_space.sample_random(n=5)
        for s in mixed_samples:
            assert 0.001 <= s["lr"] <= 0.1
            assert 0.001 <= s["lr_log"] <= 0.1

        # ── 7. Configure a random search ─────────────────────────────────

        config = SearchConfig(
            strategy="random",
            n_trials=3,  # Small for tutorial speed
            metric_to_optimize="accuracy",
            direction="maximize",
        )

        assert config.strategy == "random"
        assert config.direction == "maximize"

        base_model_spec = ModelSpec(
            model_class="sklearn.ensemble.RandomForestClassifier",
            hyperparameters={"random_state": 42},
            framework="sklearn",
        )

        eval_spec = EvalSpec(metrics=["accuracy"])

        # ── 8. Run random search ─────────────────────────────────────────

        result = await search.search(
            data=df,
            schema=schema,
            base_model_spec=base_model_spec,
            search_space=space,
            config=config,
            eval_spec=eval_spec,
            experiment_name="random_search_demo",
        )

        assert isinstance(result, SearchResult)
        assert result.strategy == "random"
        assert len(result.all_trials) == 3
        assert result.best_trial_number >= 0
        assert "accuracy" in result.best_metrics
        assert isinstance(result.best_params, dict)
        assert result.total_time_seconds > 0

        # All trials have results
        for trial in result.all_trials:
            assert isinstance(trial, TrialResult)
            assert isinstance(trial.params, dict)
            assert "accuracy" in trial.metrics
            assert trial.training_time_seconds > 0

        # ── 9. Run grid search ───────────────────────────────────────────
        # Use a small categorical-only space so grid is manageable.

        small_space = SearchSpace(
            params=[
                ParamDistribution(
                    name="max_depth",
                    type="categorical",
                    choices=[3, 5],
                ),
            ]
        )

        grid_config = SearchConfig(
            strategy="grid",
            n_trials=10,  # Grid ignores n_trials; uses full grid
            metric_to_optimize="accuracy",
            direction="maximize",
        )

        grid_result = await search.search(
            data=df,
            schema=schema,
            base_model_spec=base_model_spec,
            search_space=small_space,
            config=grid_config,
            eval_spec=eval_spec,
            experiment_name="grid_search_demo",
        )

        assert isinstance(grid_result, SearchResult)
        assert grid_result.strategy == "grid"
        assert len(grid_result.all_trials) == 2  # 2 choices for max_depth

        # ── 10. Minimization direction ───────────────────────────────────

        min_config = SearchConfig(
            strategy="random",
            n_trials=2,
            metric_to_optimize="accuracy",
            direction="minimize",
        )

        min_result = await search.search(
            data=df,
            schema=schema,
            base_model_spec=base_model_spec,
            search_space=small_space,
            config=min_config,
            eval_spec=eval_spec,
            experiment_name="min_search_demo",
        )

        # Best should be the trial with lowest accuracy
        best_acc = min_result.best_metrics.get("accuracy", 0)
        for trial in min_result.all_trials:
            assert best_acc <= trial.metrics.get("accuracy", float("inf"))

        # ── 11. Serialization round-trips ────────────────────────────────

        # ParamDistribution
        pd_dict = space.params[0].to_dict()
        pd_restored = ParamDistribution.from_dict(pd_dict)
        assert pd_restored.name == "n_estimators"
        assert pd_restored.type == "int_uniform"

        # SearchSpace
        ss_dict = space.to_dict()
        ss_restored = SearchSpace.from_dict(ss_dict)
        assert len(ss_restored.params) == len(space.params)

        # SearchConfig
        sc_dict = config.to_dict()
        sc_restored = SearchConfig.from_dict(sc_dict)
        assert sc_restored.strategy == "random"
        assert sc_restored.n_trials == 3

        # TrialResult
        tr_dict = result.all_trials[0].to_dict()
        tr_restored = TrialResult.from_dict(tr_dict)
        assert tr_restored.trial_number == result.all_trials[0].trial_number

        # SearchResult
        sr_dict = result.to_dict()
        sr_restored = SearchResult.from_dict(sr_dict)
        assert sr_restored.strategy == result.strategy
        assert len(sr_restored.all_trials) == len(result.all_trials)

        # ── 12. Edge case: invalid strategy ──────────────────────────────

        try:
            bad_config = SearchConfig(strategy="invalid_strategy")
            await search.search(
                data=df,
                schema=schema,
                base_model_spec=base_model_spec,
                search_space=space,
                config=bad_config,
                eval_spec=eval_spec,
                experiment_name="bad_strategy",
            )
            assert False, "Should raise ValueError for invalid strategy"
        except ValueError:
            pass  # Expected

    # ── 13. Clean up ─────────────────────────────────────────────────────

    await conn.close()

    print("PASS: 05-ml/08_hyperparameter_search")


asyncio.run(main())

File: textbook/rust/05-ml/src/bin/08_hyperparameter_search.rs
// Copyright 2026 Terrene Foundation
// SPDX-License-Identifier: Apache-2.0
//! Kailash SDK Textbook -- ML / Hyperparameter Search
//!
//! OBJECTIVE: Optimize model hyperparameters using grid and random search.
//! LEVEL: Intermediate
//! PARITY: Equivalent -- Python uses HyperparameterSearch;
//!         Rust uses the same search strategies with typed parameter spaces.
//! VALIDATES: GridSearch, RandomSearch, parameter space, cross-validation
//!
//! Run: cargo run -p tutorial-ml --bin 08_hyperparameter_search

use std::collections::HashMap;

fn main() {
    // ── 1. Parameter space definition ──
    // Define the hyperparameters to search over.

    let param_space = ParamSpace::new()
        .add_categorical("model", &["logistic", "svm", "random_forest"])
        .add_continuous("learning_rate", 0.001, 0.1)
        .add_discrete("max_depth", 3, 15)
        .add_discrete("n_estimators", 50, 500);

    assert_eq!(param_space.param_count(), 4);
    assert!(param_space.has_param("learning_rate"));

    // ── 2. Grid search ──
    // Exhaustively evaluates all combinations in a discretized grid.

    let grid = GridSearch::new(param_space.clone())
        .cv_folds(5)
        .metric("accuracy");

    assert_eq!(grid.cv_folds(), 5);
    assert_eq!(grid.metric(), "accuracy");

    // Grid search generates all combinations
    // For categorical: 3 values, discrete ranges produce combinatorial explosion
    // In practice, grid search is only feasible for small parameter spaces.

    // ── 3. Random search ──
    // Samples random parameter combinations -- more efficient than grid.

    let random = RandomSearch::new(param_space.clone())
        .n_trials(50)
        .cv_folds(5)
        .metric("f1_score")
        .seed(42);

    assert_eq!(random.n_trials(), 50);
    assert_eq!(random.metric(), "f1_score");

    // ── 4. Simulated search results ──
    // Each trial produces a parameter configuration and a score.

    let results = vec![
        SearchResult {
            params: HashMap::from([
                ("model".to_string(), "random_forest".to_string()),
                ("n_estimators".to_string(), "200".to_string()),
                ("max_depth".to_string(), "10".to_string()),
            ]),
            score: 0.87,
            cv_std: 0.02,
        },
        SearchResult {
            params: HashMap::from([
                ("model".to_string(), "svm".to_string()),
                ("learning_rate".to_string(), "0.01".to_string()),
            ]),
            score: 0.83,
            cv_std: 0.03,
        },
        SearchResult {
            params: HashMap::from([
                ("model".to_string(), "logistic".to_string()),
                ("learning_rate".to_string(), "0.05".to_string()),
            ]),
            score: 0.80,
            cv_std: 0.01,
        },
    ];

    // ── 5. Best result selection ──

    let best = results.iter().max_by(|a, b| {
        a.score.partial_cmp(&b.score).unwrap()
    }).unwrap();

    assert_eq!(best.params["model"], "random_forest");
    assert!((best.score - 0.87).abs() < 0.001);

    // ── 6. Cross-validation ──
    // Each search trial uses k-fold cross-validation:
    //   1. Split training data into k folds
    //   2. Train on k-1 folds, evaluate on the held-out fold
    //   3. Repeat k times, average the scores
    //   4. Report mean score and standard deviation

    // Simulate 5-fold CV scores
    let fold_scores = vec![0.85, 0.87, 0.89, 0.84, 0.86];
    let mean: f64 = fold_scores.iter().sum::<f64>() / fold_scores.len() as f64;
    let variance: f64 = fold_scores.iter().map(|s| (s - mean).powi(2)).sum::<f64>()
        / fold_scores.len() as f64;
    let std = variance.sqrt();

    assert!((mean - 0.862).abs() < 0.001);
    assert!(std < 0.02);

    // ── 7. Key concepts ──
    // - ParamSpace: defines searchable hyperparameters
    // - GridSearch: exhaustive (small spaces only)
    // - RandomSearch: sampling-based (large spaces, more efficient)
    // - Cross-validation: robust evaluation (k-fold)
    // - cv_std: standard deviation indicates stability
    // - Best result: highest mean CV score
    // - In practice, random search outperforms grid search for most problems

    println!("PASS: 05-ml/08_hyperparameter_search");
}

#[derive(Clone)]
struct ParamSpace {
    params: HashMap<String, ParamType>,
}

#[derive(Clone)]
enum ParamType {
    Categorical(Vec<String>),
    Continuous(f64, f64),
    Discrete(i64, i64),
}

impl ParamSpace {
    fn new() -> Self { Self { params: HashMap::new() } }
    fn add_categorical(mut self, name: &str, values: &[&str]) -> Self {
        self.params.insert(name.to_string(), ParamType::Categorical(values.iter().map(|s| s.to_string()).collect()));
        self
    }
    fn add_continuous(mut self, name: &str, min: f64, max: f64) -> Self {
        self.params.insert(name.to_string(), ParamType::Continuous(min, max));
        self
    }
    fn add_discrete(mut self, name: &str, min: i64, max: i64) -> Self {
        self.params.insert(name.to_string(), ParamType::Discrete(min, max));
        self
    }
    fn param_count(&self) -> usize { self.params.len() }
    fn has_param(&self, name: &str) -> bool { self.params.contains_key(name) }
}

struct GridSearch {
    _space: ParamSpace,
    cv_folds: usize,
    metric: String,
}

impl GridSearch {
    fn new(space: ParamSpace) -> Self { Self { _space: space, cv_folds: 5, metric: "accuracy".to_string() } }
    fn cv_folds(mut self, n: usize) -> Self { self.cv_folds = n; self }
    fn metric(mut self, m: &str) -> Self { self.metric = m.to_string(); self }
    fn cv_folds(&self) -> usize { self.cv_folds }
    fn metric(&self) -> &str { &self.metric }
}

struct RandomSearch {
    _space: ParamSpace,
    n_trials: usize,
    cv_folds: usize,
    metric: String,
    _seed: u64,
}

impl RandomSearch {
    fn new(space: ParamSpace) -> Self {
        Self { _space: space, n_trials: 50, cv_folds: 5, metric: "accuracy".to_string(), _seed: 42 }
    }
    fn n_trials(mut self, n: usize) -> Self { self.n_trials = n; self }
    fn cv_folds(mut self, n: usize) -> Self { self.cv_folds = n; self }
    fn metric(mut self, m: &str) -> Self { self.metric = m.to_string(); self }
    fn seed(mut self, s: u64) -> Self { self._seed = s; self }
    fn n_trials(&self) -> usize { self.n_trials }
    fn metric(&self) -> &str { &self.metric }
}

struct SearchResult {
    params: HashMap<String, String>,
    score: f64,
    cv_std: f64,
}
